pkg/dcmtime/common.go

```diff
@@ -0,0 +1,147 @@
+	"strconv"
+	"strings"
+	"time"
+)
+
+// isIncluded returns whether `check` is included in `limit`.
+//
+// Example: to test whether seconds should be included, you would:
+// isIncluded(PrecisionSeconds, [caller-passed-limit])
+func isIncluded(check PrecisionLevel, precision PrecisionLevel) bool {
+	return check <= precision
+}
+
+// truncateMilliseconds truncate nanosecond time.Time value to arbitrary precision.
+func truncateMilliseconds(
+	nanoSeconds int, precision PrecisionLevel,
+) (millisecondsStr string) {
+	milliseconds := nanoSeconds / 1000
+	millisecondsStr = fmt.Sprintf("%06d", milliseconds)
+	millisecondsStr = millisecondsStr[:6-(PrecisionFull-precision)]
+
+	return millisecondsStr
+}
+
+// Const for time.Time timezone. We don't want a timezone because we don't REALLY know
+// that a TM or DA value is UTC, so we don't want UTC to be made explicit.
+var zeroTimezone = time.FixedZone("", 0)
+
+// durationInfo holds parsed piece of DA, TM, or DT info.
+type durationInfo struct {
+	// The parsed int value
+	Value int
+	// Whether it was present in the string
+	Present bool
+	// Only used when parsing Fractal seconds: the level of precision used.
+	FractalPrecision PrecisionLevel
+}
+
+// extractDurationInfo extracts a piece of DA, TM, or DT info from a parsed regex.
+func extractDurationInfo(subMatches []string, index int, fractal bool) (
+	info durationInfo, err error,
+) {
+	if len(subMatches) <= index {
+		return info, errors.New("not enough sub-matches")
+	}
+
+	valueStr := subMatches[index]
+	// If there was no match for the specific subgroup, this value is 0.
+	if valueStr == "" {
+		valueStr = "0"
+	} else {
+		info.Present = true
+	}
+
+	// If this is a fractal seconds value, we need to pad it out to nanoseconds for go.
+	if fractal {
+		// The fractal value can be any length, with truncation implying a loss of
+		// precision, we need to add trailing zeros to the hundred-millionth place to
+		// get our nano-seconds.
+		missingPlaces := 9 - len(valueStr)
+		valueStr = valueStr + strings.Repeat("0", missingPlaces)
+		info.FractalPrecision = PrecisionFull - PrecisionLevel(missingPlaces-3)
+	}
+
+	info.Value, err = strconv.Atoi(valueStr)
+	if err != nil {
+		return info, fmt.Errorf("error parsing int: %w", err)
+	}
+
+	return info, nil
+}
+
+// updatePrecision takes in our current precision and a piece of parsed info, then
+// updates the precision based on whether the info was present.
+//
+// levelIsFull should be set to true if the level we are checking is "Full" for the
+// type of value we are parsing.
+func updatePrecision(
+	current PrecisionLevel,
+	info durationInfo,
+	infoLevel PrecisionLevel,
+	levelIsFull bool,
+) PrecisionLevel {
+	// If the info was not present, return our current level.
+	if !info.Present {
+		return current
+	}
+
+	if levelIsFull {
+		return PrecisionFull
+	}
+	return infoLevel
+}
+
+// hasMatches checks if a regex result has matches.
+func hasMatches(matches []string, original string) bool {
```

Comment:
even though these are internal utilities, reading this file as a reviewer/reader it is a bit hard to understand what the parameters mean (e.g. what's matches? Why is the 0th entry of matches special?)

Reply:
The 0th entry is the full match the regex made. Each entry in the result after index 0 is a submatch (capture group). If the 0th entry does not equal the ENTIRE original string then that means our regex got a false hit -- something that contains a valid pattern match, but also contains additional information.

Consider the following example:

```
package main

import (
	"fmt"
	"regexp"
)

// Parses dicom TM (time) value of HHMMSS.FFFFFF.
var tmRegex = regexp.MustCompile(
	`(?P<HOURS>[0-9]{2})?(?P<MINUTES>[0-9]{2})?(?P<SECONDS>[0-9]{2})?(?:\.(?P<FRACTAL>[0-9]+))?`,
)

func main() {
	stringVal := "123005.1234Hello"
	result := tmRegex.FindStringSubmatch(stringVal)

	fmt.Println(result[0])
}
```

In this example, our raw string is NOT a valid TM value, since it has "Hello" at the end, but we still get a valid hit on the regex. For this reason we need to check that the index 0 equals the original string in it's entirety, if it does not, the string value is not a valid TM value (even if it contains a valid TM value somewhere inside of it).

Reply:
I'll add some comments to explain this in the code.

Reply:
So, after some thought, I've realized that I should have brushed up on my regex again because I totally forgot that you can enforce beginning / end of line making the check that the full match and the raw value are the same unnecessary.

I've tweaked the regexes and remove the validation helper.

Also added a bunch of error test cases to make sure the regex is handling invalidating non-conformant values.